### PR TITLE
rustbuild: Add install target. #34675

### PR DESCRIPTION
**Pull request by ahmedcharles**
It just prints to the screen currently.

r? @alexcrichton 

I'm working on the next commit to actually have it install.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR! The scaffolding here looks right to me, want to fill in the target as well? It can be pretty similar to the `make install` target, which I believe just runs `make dist` followed by running the `install.sh` shell script itself.

**Comment by ahmedcharles** [comment]
Yeah, I'll work on that next.

**Comment by ahmedcharles** [comment]
@alexcrichton 

This doesn't have equivalent functionality, but it seems like a good first step and it works for at least one scenario. What would you like to have working before putting the first version on master?

File: src/bootstrap/install.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+/// Installs everything.
+pub fn install(build: &Build, stage: u32, host: &str) {
+    let prefix = build.config.prefix.as_ref().clone().map(|x| Path::new(x))
+        .unwrap_or(Path::new("/usr/local"));
+    let empty_dir = build.out.join("tmp/empty_dir");
+    t!(fs::create_dir_all(&empty_dir));
+    if build.config.docs {
+        install_sh(&build, "docs", "rust-docs", stage, host, prefix, &empty_dir);
+    }
+    install_sh(&build, "std", "rust-std", stage, host, prefix, &empty_dir);
```
Comment (alexcrichton) [inline src/bootstrap/install.rs]: In the makefiles this and the docs pass `--disable-ldconfig`, maybe that should happen here too?

File: src/bootstrap/install.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+fn install_sh(build: &Build, package: &str, name: &str, stage: u32, host: &str,
+              prefix: &Path, empty_dir: &Path) {
+    println!("Install {} stage{} ({})", package, stage, host);
+    let package_name = format!("{}-{}-{}", name, package_vers(build), host);
+
+    let mut cmd = Command::new("sh");
+    cmd.current_dir(empty_dir)
+       .arg(sanitize_sh(&tmpdir(build).join(&package_name).join("install.sh")))
+       .arg(format!("--prefix={}", sanitize_sh(&prefix)));
```
Comment (alexcrichton) [inline src/bootstrap/install.rs]: The libdir/mandir/docdir should be relatively easily scrapable from the `config.mk` I think, perhaps those could be added?

**Review by alexcrichton — COMMENTED** [review]
Thanks! Looks good to me. 

One thing I'm not entirely certain how to handle is `make install` vs `sudo make install`. Running `sudo make install` is actually super dangerous because it may corrupt your Cargo cache with a bunch of root-owned files if you have to rebuild the build system by accident.

Perhaps the makefiles for now could just have a check for if you're sudo and bail immediately if that's done? (or maybe the python script could have this check)

File: src/bootstrap/install.rs (hunk 1 above)
Comment (ahmedcharles) [inline src/bootstrap/install.rs]: Yeah, I just missed that.

**Comment by ahmedcharles** [comment]
@alexcrichton Updated with docdir, libdir and mandir.

**Comment by alexcrichton** [comment]
@bors: r+

Awesome, thanks!

**Pull request merged**